File: python/app/v1/catalog.py

from .items import Item

class Catalog:
   
    
    def __init__(self) -> None:
        self._items: dict[str, Item] = {}
    
    def add(self, item: Item) -> None:
        if item.id.value in self._items:
            raise ValueError(f"Item with ID {item.id.value} already exists in catalog")
        self._items[item.id.value] = item
    
    def get(self, item_id: str) -> Item | None:
        return self._items.get(item_id)
    
    def contains(self, item_id: str) -> bool:
        return item_id in self._items

File: rust/src/v1/catalog.rs
use std::collections::HashMap;
use crate::v1::items::Item;

pub struct Catalog {
    items: HashMap<String, Box<dyn Item>>,
}

impl Catalog {
    pub fn new() -> Self {
        Catalog {
            items: HashMap::new(),
        }
    }
    
    pub fn add(&mut self, item: Box<dyn Item>) -> Result<(), String> {
        let id = item.id().to_string();
        
        if self.items.contains_key(&id) {
            return Err(format!("Item with ID {} already exists in catalog", id));
        }
        
        self.items.insert(id, item);
        Ok(())
    }
    
    pub fn get(&self, item_id: &str) -> Option<&dyn Item> {
        self.items.get(item_id).map(|boxed| boxed.as_ref())
    }
    
    pub fn contains(&self, item_id: &str) -> bool {
        self.items.contains_key(item_id)
    }
    
    pub fn details_for(&self, item_ids: &[String]) -> Vec<(String, String, u32)> {
        item_ids
            .iter()
            .filter_map(|id| {
                self.get(id).map(|item| {
                    (id.clone(), item.title().to_string(), item.days_allowed())
                })
            })
            .collect()
    }
}